module_checksum_upgrade/models/module.py

```diff
@@ -0,0 +1,134 @@
+PARAM_EXCLUDED_EXTENSIONS = \
+    'module_checksum_upgrade.checksum_excluded_extensions'
+
+
+class Module(models.Model):
+    _inherit = 'ir.module.module'
+
+    @api.multi
+    def _get_checksum_dir(self):
+        self.ensure_one()
+
+        excluded_extensions = self.env["ir.config_parameter"].get_param(
+            PARAM_EXCLUDED_EXTENSIONS,
+            'pyc,pyo',
+        ).split(",")
+
+        module_path = get_module_path(self.name)
+        if module_path and os.path.isdir(module_path):
+            checksum_dir = dirhash(
+                module_path,
+                'sha1',
+                excluded_extensions=excluded_extensions,
+            )
+        else:
+            checksum_dir = False
+
+        return checksum_dir
+
+    @api.model
+    def _get_saved_checksums(self):
+        Icp = self.env['ir.config_parameter']
+        return json.loads(Icp.get_param(PARAM_INSTALLED_CHECKSUMS, '{}'))
+
+    @api.model
+    def _save_checksums(self, checksums):
+        Icp = self.env['ir.config_parameter']
+        Icp.set_param(PARAM_INSTALLED_CHECKSUMS, json.dumps(checksums))
+
+    @api.model
+    def _save_installed_checksums(self):
+        checksums = {}
+        installed_modules = self.search([('state', '=', 'installed')])
+        for module in installed_modules:
+            checksums[module.name] = module._get_checksum_dir()
+        self._save_checksums(checksums)
+
+    @api.model
+    def _get_modules_partially_installed(self):
+        return self.search([
+            '!',
+            ('state', 'in', ('installed', 'uninstalled', 'uninstallable')),
+        ])
+
+    @api.model
+    def _get_modules_with_changed_checksum(self):
+        saved_checksums = self._get_saved_checksums()
+        installed_modules = self.search([('state', '=', 'installed')])
+        return installed_modules.filtered(
+            lambda r: r._get_checksum_dir() != saved_checksums.get(r.name),
+        )
+
+    @api.model
+    def upgrade_changed_checksum(self):
+        """ Run an upgrade of the database, upgrading only changed modules.
+
+        Installed modules for which the checksum has changed since the
+        last successful run of this method are marked "to upgrade",
+        then the normal Odoo scheduled upgrade process
+        is launched.
+
+        If there is no module with a changed checksum, and no module in state
+        other than installed, uninstalled, uninstallable, this method does
+        nothing, otherwise the normal Odoo upgrade process is launched.
+
+        After a successful upgrade, the checksums of installed modules are
+        saved.
+
+        In case of error during the upgrade, an exception is raised.
+        If any module remains to upgrade or to uninstall after the upgrade
+        process, this method returns False. It returns True only in case
+        of complete success.
+        """
+        _logger.info("Upgrade starting...")
+
+        _logger.info("Updating modules list...")
```

Comment:
2 logs for the same thing?

Reply:
These are not the same: the first is to log the start of the whole checksum upgrade process, then each step is logged (updating module list, marking to upgrade, etc)